source/common/ssl/tls_certificate_config_impl.cc

```diff
@@ -4,13 +4,27 @@
 #include "envoy/extensions/transport_sockets/tls/v3/cert.pb.h"
 #include "envoy/server/transport_socket_config.h"
 
+#include "common/common/base64.h"
 #include "common/common/empty_string.h"
 #include "common/common/fmt.h"
 #include "common/config/datasource.h"
 
 namespace Envoy {
 namespace Ssl {
 
+namespace {
+std::vector<uint8_t> readDerEncodedOcspStaple(const envoy::config::core::v3::DataSource& source,
+                                              Api::Api& api) {
+  std::string staple = Config::DataSource::read(source, true, api);
+  if (source.specifier_case() ==
+      envoy::config::core::v3::DataSource::SpecifierCase::kInlineString) {
+    staple = Base64::decode(staple);
```

Comment:
There are a bunch of other fields for tls cert related stuff (private key, password, etc) that use a DataSource but don't have a base64 decode for the string case. Why is this one different? Maybe add comment?

Reply:
I think for the other data types, boringSSL will accept with a PEM-encoded DER blob (so base64 with `-----BEGIN FOO-----` and `------END FOO-----` labels) or binary DER.

So the `inline_string` can contain that PEM-encoded blob without issue and boringSSL silently handles the base64 decoding (and maybe validates the label, I'm not sure) for the password, cert, etc. There isn't a PEM format for OCSP responses that I can find, and boringSSL definitely expects a binary-encoded DER. It literally does nothing with that binary blob except send it to clients that ask for stapling.

For `inline_bytes` the protobuf yaml/json decoding expects base64. For `inline_string` I tried hex-encoding the bytes and passing them in `inline_string`, but the yaml decoder (or protobuf) mangles the bytes. I think it's trying to make a valid UTF-8 string out of my bytes. So we require base64 for inline_string. I'll update the proto docs to note this.

Reply:
In that case probably just document OCSP shouldn't be specified via `inline_string` but use `inline_bytes` instead. Any reason to support base64 encoded `inline_string` for a new field?

Reply:
I think operators might wish to pass around OCSP responses in base64 format even though there isn't an official PEM encoding. This gives them a way to do that. The distinction is murky in YAML/JSON since inline_bytes has to be a base64-string there, but from a control plane server there is a distinction.

Reply:
Why a control plane couldn't fill `inline_bytes` but a base-64 encoded `inline-string`?